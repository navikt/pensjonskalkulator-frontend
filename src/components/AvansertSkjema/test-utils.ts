export function getPreviousMonth(): string {
  const today = new Date()
  const currentMonth = parseInt(
    today.toLocaleDateString('en-CA').slice(5, 7),
    10
  ) // Extracts the month from the date string
  const previousMonth = currentMonth - 1
  return previousMonth.toString()
}
