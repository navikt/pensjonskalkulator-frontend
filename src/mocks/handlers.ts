import { delay, http, HttpResponse } from 'msw'

import { API_PATH, HOST_BASEURL } from '@/paths'

import ansattIdResponse from './data/ansatt-id.json' with { type: 'json' }
import ekskludertStatusResponse from './data/ekskludert-status.json' with { type: 'json' }
import inntektResponse from './data/inntekt.json' with { type: 'json' }
import omstillingsstoenadOgGjenlevendeResponse from './data/omstillingsstoenad-og-gjenlevende.json' with { type: 'json' }
import personResponse from './data/person.json' with { type: 'json' }
import tidligstMuligHeltUttakResponse from './data/tidligstMuligHeltUttak.json' with { type: 'json' }
import tpoMedlemskapResponse from './data/tpo-medlemskap.json' with { type: 'json' }
import ufoeregradResponse from './data/ufoeregrad.json' with { type: 'json' }
import disableSpraakvelgerToggleResponse from './data/unleash-disable-spraakvelger.json' with { type: 'json' }
import highchartsAccessibilityPluginToggleResponse from './data/unleash-enable-highcharts-accessibility-plugin.json' with { type: 'json' }
import enableUtlandPluginToggleResponse from './data/unleash-enable-utland.json' with { type: 'json' }
import enableUtvidetSimuleringsresultatPluginToggleResponse from './data/unleash-utvidet-simuleringsresultat.json' with { type: 'json' }

const TEST_DELAY = process.env.NODE_ENV === 'test' ? 0 : 30

export const getHandlers = (baseUrl: string = API_PATH) => [
  http.get(`${HOST_BASEURL}/oauth2/session`, async () => {
    await delay(1500)
    return HttpResponse.json({ data: 'OK' })
  }),

  http.get(`${baseUrl}/inntekt`, async () => {
    await delay(TEST_DELAY)
    return HttpResponse.json(inntektResponse)
  }),

  http.get(`${baseUrl}/v2/ekskludert`, async () => {
    await delay(TEST_DELAY)
    return HttpResponse.json(ekskludertStatusResponse)
  }),

  http.get(
    `${baseUrl}/v1/loepende-omstillingsstoenad-eller-gjenlevendeytelse`,
    async () => {
      await delay(TEST_DELAY)
      return HttpResponse.json(omstillingsstoenadOgGjenlevendeResponse)
    }
  ),

  http.get(`${baseUrl}/v1/ufoeregrad`, async () => {
    await delay(TEST_DELAY)
    return HttpResponse.json(ufoeregradResponse)
  }),

  http.get(`${baseUrl}/v2/person`, async ({ request }) => {
    await delay(TEST_DELAY)
    if (request.headers.get('fnr') === '40100000000') {
      return HttpResponse.json({}, { status: 401 })
    }
    if (request.headers.get('fnr') === '40400000000') {
      return HttpResponse.json({}, { status: 404 })
    }

    return HttpResponse.json(personResponse)
  }),

  http.get(`${baseUrl}/v1/ansatt-id`, async () => {
    await delay(TEST_DELAY)
    return HttpResponse.json(ansattIdResponse)
  }),

  http.get(`${baseUrl}/v1/tpo-medlemskap`, async () => {
    await delay(TEST_DELAY)
    return HttpResponse.json(tpoMedlemskapResponse)
  }),

  http.post(`${baseUrl}/v1/tidligste-hel-uttaksalder`, async () => {
    await delay(TEST_DELAY)
    return HttpResponse.json(tidligstMuligHeltUttakResponse)
  }),

  http.post(`${baseUrl}/v2/pensjonsavtaler`, async ({ request }) => {
    await delay(TEST_DELAY)
    const body = await request.json()
    const aar = (body as PensjonsavtalerRequestBody).uttaksperioder[0]
      ?.startAlder.aar
    const data = await import(`./data/pensjonsavtaler/${aar}.json`)
    return HttpResponse.json(data)
  }),

  http.post(`${baseUrl}/v6/alderspensjon/simulering`, async ({ request }) => {
    await delay(TEST_DELAY)
    const body = await request.json()
    const aar = (body as AlderspensjonRequestBody).heltUttak.uttaksalder.aar
    const data = await import(`./data/alderspensjon/${aar}.json`)
    const mergedData = JSON.parse(JSON.stringify(data.default))
    let afpPrivat: Pensjonsberegning[] = []
    let afpOffentlig: Pensjonsberegning[] = []
    if (
      (body as AlderspensjonRequestBody).simuleringstype ===
      'ALDERSPENSJON_MED_AFP_PRIVAT'
    ) {
      const afpPrivatData = JSON.parse(
        JSON.stringify(await import(`./data/afp-privat/${aar}.json`))
      )
      afpPrivat = [...afpPrivatData.default.afpPrivat]
    }
    if (
      (body as AlderspensjonRequestBody).simuleringstype ===
      'ALDERSPENSJON_MED_AFP_OFFENTLIG_LIVSVARIG'
    ) {
      const afpOffentligData = JSON.parse(
        JSON.stringify(await import(`./data/afp-offentlig.json`))
      )
      if (afpOffentligData.default.afpOffentlig) {
        afpOffentlig = [
          {
            ...afpOffentligData.default.afpOffentlig[0],
            alder: aar,
          },
        ]
      }
    }

    return HttpResponse.json({
      ...mergedData,
      afpPrivat: afpPrivat,
      afpOffentlig: afpOffentlig,
    })
  }),

  http.post(`${baseUrl}/v1/encrypt`, async () => {
    await delay(TEST_DELAY)
    return HttpResponse.text('this-is-just-jibbrish-encrypted-fnr')
  }),

  http.get(
    `${baseUrl}/feature/pensjonskalkulator.disable-spraakvelger`,
    async () => {
      await delay(TEST_DELAY)
      return HttpResponse.json(disableSpraakvelgerToggleResponse)
    }
  ),

  http.get(
    `${baseUrl}/feature/pensjonskalkulator.enable-highcharts-accessibility-plugin`,
    async () => {
      await delay(TEST_DELAY)
      return HttpResponse.json(highchartsAccessibilityPluginToggleResponse)
    }
  ),

  http.get(`${baseUrl}/feature/pensjonskalkulator.enable-utland`, async () => {
    await delay(TEST_DELAY)
    return HttpResponse.json(enableUtlandPluginToggleResponse)
  }),

  http.get(
    `${baseUrl}/feature/pensjonskalkulator.utvidet-simuleringsresultat`,
    async () => {
      await delay(TEST_DELAY)
      return HttpResponse.json(
        enableUtvidetSimuleringsresultatPluginToggleResponse
      )
    }
  ),

  http.post('http://localhost:12347/collect', async ({ request }) => {
    await request.json()
    return HttpResponse.json({ data: 'OK' })
  }),
]
