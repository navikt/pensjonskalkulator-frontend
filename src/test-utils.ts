import renderer from 'react-test-renderer'

import { render } from '@testing-library/react'

export function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON()
  expect(result).toBeDefined()
  expect(result).not.toBeInstanceOf(Array)
  return result as renderer.ReactTestRendererJSON
}

const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => children,
    ...options,
  })

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
// override render export
export { customRender as render }
