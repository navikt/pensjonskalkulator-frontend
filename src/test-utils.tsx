import React, { PropsWithChildren } from 'react'
import { IntlProvider } from 'react-intl'
import { Provider } from 'react-redux'
import { MemoryRouter } from 'react-router-dom'

import { PreloadedState, createListenerMiddleware } from '@reduxjs/toolkit'
import { render, RenderOptions } from '@testing-library/react'

import { getTranslation_test } from '@/utils/__tests__/test-translations'

import { createUttaksalderListener } from './state/listeners/uttaksalderListener'
import {
  setupStore,
  RootState,
  AppStore,
  AppStartListening,
} from './state/store'
import { getTranslation_nb } from './translations/nb'

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
  hasRouter?: boolean
}

export const swallowErrors = (testFn: () => void) => {
  const error = console.error
  console.error = () => {}
  testFn()
  console.error = error
}

export const swallowErrorsAsync = async (testFn: () => Promise<void>) => {
  const cache = console.error
  console.error = () => {}
  await testFn()
  console.error = cache
}

function generateMockedTranslations() {
  const nbTranslations: Record<string, string> = getTranslation_nb()
  const testTranslations: Record<string, string> = getTranslation_test()
  const translationsInput = { ...nbTranslations, ...testTranslations }
  const translations: Record<string, string> = {}

  for (const key in translationsInput) {
    if (
      /<(?=.*? .*?\/ ?>|br|hr|input|!--|wbr)[a-z]+.*?>|<([a-z]+).*?<\/\1>|{[^}]+}/i.test(
        translationsInput[key]
      )
    ) {
      // for html or chunks keys: results in 'my_key' : 'my_key with some <html> or {chunk}'
      translations[key] = translationsInput[key]
    } else {
      // results in 'my_key' : 'my_key'
      translations[key] = key
    }
  }
  return translations
}

// Return an object with the store and all of RTL's query functions
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState, true),
    hasRouter = true,
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<unknown>): JSX.Element {
    return (
      <Provider store={store}>
        <IntlProvider locale="nb" messages={generateMockedTranslations()}>
          {hasRouter ? <MemoryRouter>{children}</MemoryRouter> : children}
        </IntlProvider>
      </Provider>
    )
  }

  const listenerMiddleware = createListenerMiddleware()
  createUttaksalderListener(
    listenerMiddleware.startListening as AppStartListening
  )

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'

export { renderWithProviders as render }
