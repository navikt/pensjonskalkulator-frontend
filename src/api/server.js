// const express = require('express')
// const liveness = require('../src/__mocks__/data/liveness.json')
// const pensjonsberegning = require('../src/__mocks__/data/pensjonsberegning.json')

// const app = express()
// const port = 8088

// app.get('/api/status', (request, response) => {
//   response.send(liveness)
// })

// app.get('/api/pensjonsberegning', async (request, response) => {
//   response.send(pensjonsberegning)
// })

// app.listen(process.env.PORT || port, () => {
//   console.info('proxy server is running on port', process.env.PORT || port)
// })

// import { rest } from 'msw'
// import { setupServer } from 'msw/node'

import { setupServer } from 'msw/node'
import { handlers } from './handlers.js'

export const server = setupServer(...handlers)
server.listen({ onUnhandledRequest: 'error' })
