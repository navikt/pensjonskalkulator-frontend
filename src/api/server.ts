import { setupServer } from 'msw/node'
import { getHandlers } from './handlers.js'
import { rest } from 'msw'

export const API_TARGET = 'http://localhost:8088'
export const API_PATH = '/pensjon/kalkulator/api'

const handlers = getHandlers(`${API_TARGET}${API_PATH}`)
export const server = setupServer(...handlers)

server.listen({ onUnhandledRequest: 'error' })

type MockResponseOptions = {
  status?: number
  json?: any
}

export const mockResponse = (
  path: string,
  options: MockResponseOptions = {
    status: 200,
    json: 'OK',
  }
) => {
  server.use(
    rest.get(`${API_TARGET}${API_PATH}${path}`, (req, res, ctx) => {
      return res(ctx.status(options.status ?? 200), ctx.json(options.json))
    })
  )
}

export const mockErrorResponse = (
  path: string,
  options: MockResponseOptions = {
    status: 500,
    json: "Beep boop I'm an error!",
  }
) => {
  mockResponse(path, {
    ...options,
    status: options.status ?? 500,
  })
}
