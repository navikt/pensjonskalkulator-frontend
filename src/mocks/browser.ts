import { setupWorker } from 'msw/browser'

import { getHandlers } from './handlers'

const handlers = getHandlers()
export const worker = setupWorker(...handlers)
