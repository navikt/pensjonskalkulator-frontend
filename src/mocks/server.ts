import { rest } from 'msw'
import { setupServer } from 'msw/node'

import { API_BASEURL } from '@/paths'

import { getHandlers } from './handlers'

const handlers = getHandlers(API_BASEURL)
export const server = setupServer(...handlers)

type MockResponseOptions = {
  status?: number
  json?: ReturnType<typeof JSON.parse>
  text?: string
  method?: 'post' | 'get'
  baseUrl?: string
}

const defaultResponseOptions: Required<MockResponseOptions> = {
  status: 200,
  json: { data: 'OK' },
  method: 'get',
  baseUrl: `${API_BASEURL}`,
  text: '',
}

export const mockResponse = (
  path: string,
  inputOptions: MockResponseOptions = {}
) => {
  const options = { ...defaultResponseOptions, ...inputOptions }
  const parsedPath = `${options.baseUrl}${path}`

  server.use(
    rest[options.method](parsedPath, (_req, res, ctx) => {
      return res(
        ctx.status(options.status),
        options.text ? ctx.text(options.text) : ctx.json(options.json)
      )
    })
  )
}

const defaultErrorResponseOptions: Required<MockResponseOptions> = {
  status: 500,
  json: { data: "Beep boop I'm an error!" },
  text: '',
  method: 'get',
  baseUrl: `${API_BASEURL}`,
}

export const mockErrorResponse = (
  path: string,
  inputOptions: MockResponseOptions = {}
) => {
  const options = {
    ...defaultErrorResponseOptions,
    ...inputOptions,
  }
  mockResponse(path, options)
}
