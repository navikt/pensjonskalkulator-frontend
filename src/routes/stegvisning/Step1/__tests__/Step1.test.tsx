import * as ReactRouterUtils from 'react-router'

import { describe, it, vi } from 'vitest'

import { Step1 } from '..'
import { mockErrorResponse } from '@/mocks/server'
import { fireEvent, render, screen, waitFor } from '@/test-utils'

describe('Step 1', () => {
  it('henter personopplysninger og viser hilsen med fornavnet til brukeren', async () => {
    render(<Step1 />)
    await waitFor(() => {
      expect(
        screen.getByText('stegvisning.stegvisning.start.title Aprikos!')
      ).toBeVisible()
    })
  })

  it('render ingenting når henting av personopplysninger feiler', async () => {
    mockErrorResponse('/person')
    const { asFragment } = render(<Step1 />)

    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot()
    })
  })

  it('sender videre til steg 2 når brukeren klikker på Neste', async () => {
    const navigateMock = vi.fn()
    vi.spyOn(ReactRouterUtils, 'useNavigate').mockImplementation(
      () => navigateMock
    )
    render(<Step1 />)
    await waitFor(() => {
      fireEvent.click(screen.getByText('stegvisning.stegvisning.start.start'))
      expect(navigateMock).toHaveBeenCalledWith('/samtykke')
    })
  })

  it('redirigerer til landingssiden når brukeren klikker på Avbryt', async () => {
    const navigateMock = vi.fn()
    vi.spyOn(ReactRouterUtils, 'useNavigate').mockImplementation(
      () => navigateMock
    )
    render(<Step1 />)
    await waitFor(() => {
      fireEvent.click(screen.getByText('stegvisning.avbryt'))
      expect(navigateMock).toHaveBeenCalledWith('/')
    })
  })
})
