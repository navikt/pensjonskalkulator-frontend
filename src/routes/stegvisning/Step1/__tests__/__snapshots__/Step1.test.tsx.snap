// Vitest Snapshot v1

exports[`Step 1 > render ingenting nÃ¥r henting av personopplysninger feiler 1`] = `<DocumentFragment />`;
